File: mappa antica/mappa.cpp
#include <bits/stdc++.h>

using namespace std;

void mappa(int x, int y, int N, vector<vector<char> >& map, vector<vector<int> >& dist, int distance){
	
	dist[y][x] = distance;
	
	char ast = '*';
	
	// up, down, right, left
	if (y < N - 1){
		if (map[y + 1][x] == ast){
			if (dist[y + 1][x] > distance + 1){
				mappa(x, y + 1, N, map, dist, distance + 1);
			}
		}
	}
	if (y > 0){
		if (map[y - 1][x] == ast){
			if (dist[y - 1][x] > distance + 1){
				mappa(x, y - 1, N, map, dist, distance + 1);
			}
		}
	}
	if (x < N - 1){
		if (map[y][x + 1] == ast){
			if (dist[y][x + 1] > distance + 1){
				mappa(x + 1, y, N, map, dist, distance + 1);
			}
		}
	}
	if (x > 0){
		if (map[y][x - 1] == ast){
			if (dist[y][x - 1] > distance + 1){
				mappa(x - 1, y, N, map, dist, distance + 1);
			}
		}
	}
	
	// up-right, up-left, down-left, down-right
	
	if (x < N - 1 && y < N - 1){
		if (map[y + 1][x + 1] == ast){
			if (dist[y + 1][x + 1] > distance + 1){
				mappa(x + 1, y + 1, N, map, dist, distance + 1);
			}
		}
	}
	if (x < N - 1 && y > 0){
		if (map[y - 1][x + 1] == ast){
			if (dist[y - 1][x + 1] > distance + 1){
				mappa(x + 1, y - 1, N, map, dist, distance + 1);
			}
		}
	}
	if (x > 0 && y > 0){
		if (map[y - 1][x - 1] == ast){
			if (dist[y - 1][x - 1] > distance + 1){
				mappa(x - 1, y - 1, N, map, dist, distance + 1);
			}
		}
	}
	if (x > 0 && y < N - 1){
		if (map[y + 1][x - 1] == ast){
			if (dist[y + 1][x - 1] > distance + 1){
				mappa(x - 1, y + 1, N, map, dist, distance + 1);
			}
		}
	}
	
	
	
}

int main(){
	
	
//	freopen("input.txt", "r", stdin);
//	freopen("output.txt", "w", stdout);
	
	int N;
	
	cin >> N;
	
	vector<vector<char> > map(N, vector<char>(N));
	
	for (int i = 0; i < N; i++){
		for (int j = 0; j < N; j++){
			cin >> map[i][j];
		}
	}
	
	vector<vector<int> > dist(N, vector<int>(N, 1e9));
	
	mappa(0, 0, N, map, dist, 0);
	
	cout << dist[N - 1][N - 1] + 1 << "\n";
	
	
	return 0;
}

File: mappa antica/mappa.py
#Scrivi qui il tuo codice

import sys
sys.stdin = open("input.txt", "r")
sys.stdout = open("output.txt", "w")




class cell:
    def __init__(self, x, y, cella):
        self.x = x
        self.y = y
        self.cella = cella
        self.distance = 100 * 100
    
    def calculate(self, mappa, distance):
        
        self.distance = distance
        
        if self.x > 0:
            if mappa[self.y][self.x - 1].distance > self.distance + 1:
                if mappa[self.y][self.x - 1].cella == '*':
                    mappa[self.y][self.x - 1].calculate(mappa, self.distance + 1)
        
        if self.x < N - 1:
            if mappa[self.y][self.x + 1].distance > self.distance + 1:
                if mappa[self.y][self.x + 1].cella == '*':
                    mappa[self.y][self.x + 1].calculate(mappa, self.distance + 1)
        
        
        if self.y > 0:
            if mappa[self.y - 1][self.x].distance > self.distance + 1:
                if mappa[self.y - 1][self.x].cella == '*':
                    mappa[self.y - 1][self.x].calculate(mappa, self.distance + 1)
        
        if self.y < N - 1:
            if mappa[self.y + 1][self.x].distance > self.distance + 1:
                if mappa[self.y + 1][self.x].cella == '*':
                    mappa[self.y + 1][self.x].calculate(mappa, self.distance + 1)
        
        
        
        if self.x > 0 and self.y > 0:
            if mappa[self.y - 1][self.x - 1].distance > self.distance + 1:
                if mappa[self.y - 1][self.x - 1].cella == '*':
                    mappa[self.y - 1][self.x - 1].calculate(mappa, self.distance + 1)
        
        if self.x < N - 1 and self.y > 0:
            if mappa[self.y - 1][self.x + 1].distance > self.distance + 1:
                if mappa[self.y - 1][self.x + 1].cella == '*':
                    mappa[self.y - 1][self.x + 1].calculate(mappa, self.distance + 1)
        
        if self.x > 0 and self.y < N - 1:
            if mappa[self.y + 1][self.x - 1].distance > self.distance + 1:
                if mappa[self.y + 1][self.x - 1].cella == '*':
                    mappa[self.y + 1][self.x - 1].calculate(mappa, self.distance + 1)
        
        if self.x < N - 1 and self.y < N - 1:
            if mappa[self.y + 1][self.x + 1].distance > self.distance + 1:
                if mappa[self.y + 1][self.x + 1].cella == '*':
                    mappa[self.y + 1][self.x + 1].calculate(mappa, self.distance + 1)
        
        




N = int(input())

mappa = [[''] * N] * N

for i in range(N):
    mappa[i] = [char for char in str(input().strip())]
    for j in range(len(mappa[i])):
        mappa[i][j] = cell(j, i, mappa[i][j])
    

mappa[0][0].calculate(mappa, 1)

print(mappa[N-1][N-1].distance)



#for i in mappa:
    #for j in i:
        #if j.distance == 100*100:
            #j.distance = -1
        #print(j.x, j.y, j.distance, " ", end='')
    #print()
